src/SurfaceFluxes.jl

```diff
@@ -54,186 +65,445 @@ Surface flux conditions, returned from `surface_conditions`.
-            ϕ = x_tup[i]
-            transport = i - 1 == 1 ? UF.MomentumTransport() : UF.HeatTransport()
-            F_i =
-                ϕ - compute_physical_scale(
-                    uf,
-                    z_in,
-                    length(nt.z_0) > 1 ? z_0[i - 1] : z_0,
-                    x_in[i - 1],
-                    x_s[i - 1],
-                    transport,
-                    scheme,
-                )
-        end
-        F_i
-    end
-    F .= F_nt
+struct Fluxes{VI, VS, AUF, FT} <: AbstractSurfaceConditions
+    vals_in::VI
+    vals_sfc::VS
+    shf::FT
+    lhf::FT
+    uf::AUF
+    z0m::FT
+    z0b::FT
+    L_MO_init::FT
 end
 
-"""
-    surface_conditions
-Surface conditions given
- - `MO_param_guess` initial guess for solution (`L_MO, u_star, θ_star, ϕ_star, ...`)
- - `x_in` Inner values for state variable array `x`. They correspond to volume averages in FV and
-    nodal point values in DG
- - `x_s` surface values for state variable array `x`
- - `z_0` roughness lengths for state variable array `x`
- - `θ_scale` virtual dry static energy (i.e., basic potential temperature)
- - `z_in` Input height for the similarity functions. It is Δz for FV and the height
-    of the first nodal point for DG
- - `wθ_flux_star` potential temperature flux (optional)
-  - `universal_func` family of flux-gradient universal functions used (optional)
+struct FluxesAndFrictionVelocity{VI, VS, AUF, FT} <:
+       AbstractSurfaceConditions
+    vals_in::VI
+    vals_sfc::VS
+    shf::FT
+    lhf::FT
+    ustar::FT
+    uf::AUF
+    z0m::FT
+    z0b::FT
+end
 
-If `wθ_flux_star` is not given, then it is computed by iteration
-of equations 3, 17, and 18 in Nishizawa2018.
-"""
-function surface_conditions(
-    param_set::APS,
-    MO_param_guess::AbstractVector,
-    x_in::AbstractVector,
-    x_s::AbstractVector,
-    z_0::Union{AbstractVector, FT},
-    θ_scale::FT,
-    z_in::FT,
-    scheme,
-    wθ_flux_star::Union{Nothing, FT} = nothing,
-    universal_func::Union{Nothing, F} = UF.Businger,
-    sol_type::NS.SolutionType = NS.CompactSolution(),
-    tol::NS.AbstractTolerance = NS.ResidualTolerance{FT}(sqrt(eps(FT))),
-    maxiter::Int = 10_000,
-) where {FT <: AbstractFloat, APS, F}
-
-    n_vars = length(MO_param_guess) - 1
-    @assert length(MO_param_guess) == n_vars + 1
-    @assert length(x_in) == n_vars
-    @assert length(x_s) == n_vars
-    local sol
+struct Coefficients{VI, VS, AUF, FT} <: AbstractSurfaceConditions
+    vals_in::VI
+    vals_sfc::VS
+    Cd::FT
+    Ch::FT
+    uf::AUF
+    z0m::FT
+    z0b::FT
+end
+
+struct ValuesOnly{VI, VS, AUF, FT} <: AbstractSurfaceConditions
+    vals_in::VI
+    vals_sfc::VS
+    uf::AUF
+    z0m::FT
+    z0b::FT
+    L_MO_init::FT
+end
+
+struct ValuesSurface{FT} <: AbstractSurfaceLocation
+    ts::TD.ThermodynamicState
+    u::AbstractVector
+    z::FT
+end
+struct ValuesInterior{FT} <: AbstractSurfaceLocation
+    ts::TD.ThermodynamicState
+    u::AbstractVector
+    z::FT
+end
```

Comment:
```suggestion
struct ValuesSurface{FT, A} <: AbstractSurfaceLocation
    ts::TD.ThermodynamicState
    u::A
    z::FT
end
struct ValuesInterior{FT, A} <: AbstractSurfaceLocation
    ts::TD.ThermodynamicState
    u::A
    z::FT
end
```
using `::AbstractVector` for struct properties will not work on the gpu.